chill-scala/src/test/scala/com/twitter/chill/ClosureCleanerSpec.scala

```diff
@@ -16,59 +16,227 @@ limitations under the License.
+
+      val closure1 = (i: Int) => {
+        Option(i).map { x =>
+          x + someSerializableValue
+        }
+      }
+      val closure2 = (j: Int) => {
+        Option(j).map { x =>
+          x + someSerializableMethod()
+        }
+      }
+      val closure3 = (m: Int) => {
+        Option(m).foreach { x =>
+          Option(x).foreach { y =>
+            Option(y).foreach { z =>
+              someSerializableValue
+            }
+          }
+        }
       }
 
-      ClosureCleaner(myFun)
-      myFun(1) should equal(2)
-      myFun(2) should equal(4)
+      serializableFn(closure1, before = false, after = true)
+      serializableFn(closure2, before = false, after = true)
+      serializableFn(closure3, before = false, after = true)
+    }
 
-      case class Test(x: Int)
-      val t = Test(3)
-      ClosureCleaner(t)
-      t should equal(Test(3))
+    "clean basic nested closures (LMF)" in {
+      assume(ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (i: Int) => {
+        Option(i).map { x =>
+          x + someSerializableValue
+        }
+      }
+      val closure2 = (j: Int) => {
+        Option(j).map { x =>
+          x + someSerializableMethod()
+        }
+      }
+      val closure3 = (m: Int) => {
+        Option(m).foreach { x =>
+          Option(x).foreach { y =>
+            Option(y).foreach { z =>
+              someSerializableValue
+            }
+          }
+        }
+      }
+
+      serializableFn(closure1, before = false, after = false)
+      serializableFn(closure2, before = false, after = false)
+      serializableFn(closure3, before = false, after = false)
     }
 
-    "handle outers with constructors" in {
+    "clean outers with constructors" in {
       class Test(x: String) {
-        val l = x.size
-        def rev(y: String) = (x + y).size
+        val l: Int = x.length
+        def rev(y: String): Int = (x + y).length
       }
 
       val t = new Test("you all everybody")
-      val fn = t.rev _
-      //debug(fn)
-      //println(ClosureCleaner.getOutersOf(fn))
-      ClosureCleaner(fn)
-      fn("hey") should equal(20)
+      val fn: String => Int = t.rev
+      serializableFn(fn, before = false, after = false)
     }
-    "Handle functions in traits" in {
-      val fn = BaseFns2.timesByMult
-      ClosureCleaner(fn)
-      fn(10) should equal(50)
+
+    "clean outers with constructors only non-LMF" in {
+      assume(!ClosureCleanerSpec.supportsSerializedLambda)
+      // currently this doesn't work with 2.12.x; SerializedLambda does not provide an outer ref.
+      // need to find another way...
+      serializableFn(new TestClass().fn, before = false, after = true)
     }
-    "Handle captured vals" in {
-      val answer = 42
-      val fn = { x: Int =>
-        answer * x
-      }
-      ClosureCleaner(fn)
-      fn(10) should equal(420)
+
+    "clean outers with objects" in {
+      serializableFn(TestObject.fn, before = true, after = true)
```

Comment:
how do we know this is doing anything if `before = true`? Can we find a case with an object where `before = false`?

Reply:
yeah good point; in this case I guess the thought that if something is serializable before after applying `CloserCleaner` it should still be serializable. I tried to find a case where it fails but I couldn't come up with one.